### ISSUE TITLE
Summary of Script window changes requested

### ISSUE THREAD

**Issue by lloyddewit** [africanmathsinitiative/R-Instat#7989]
This issue summarizes all the currently open changes requested that are related to the Script window:

- [ ] add tabbed windows
- [ ] tidy up file open/save/save-as functionality
- [ ] add reformat option to tidy indentation and line breaks (including splitting long lines)
- [ ] make 'ctrl-enter' execute the R statement at the cursor (rather than just the line), this should fix the current bugs with multi-line statements 
- [ ] display cursor row/column at bottom of editor
- [ ] add auto-complete
- [ ] find/replace dialog
- [ ] ensure commands on R-Instat 'Edit' menu work as expected in script window
- [ ] ensure controls translate to French and other languages
- [ ] update help files
- [ ] test on Unix
- [ ] What is the rationale/aim of the Edit>Script dialog. Should we keep it separate from the normal script window? Or should we add the Edit/Script dialog features to the Script window?
- [ ] Should all the script R-Instat displays be displayed consistently (same font, syntax highlighting and structure):
  - Script dialog
  - Output window
  - Log window
  - Edit/Script dialog
  - message boxes, help?
  - is script displayed anywhere else in R-Instat?
To achieve the above, we could create a reusable/configurable editor window that could be used in some or all of the above (text would be set to read-only for most windows/dialogs).

**Comparison with RStudio editor**

If anyone wants to compare against RStudio, then the RStudio editor functionality is described below.

https://support.rstudio.com/hc/en-us/articles/200484448-Editing-and-Executing-Code-in-the-RStudio-IDE
https://support.rstudio.com/hc/en-us/articles/200710523-Navigating-Code-in-the-RStudio-IDE
https://support.rstudio.com/hc/en-us/articles/205273297-Code-Completion-in-the-RStudio-IDE
https://support.rstudio.com/hc/en-us/articles/200484568-Code-Folding-and-Sections-in-the-RStudio-IDE

**Related issues**:

All related issues should be tagged with the `log_script window` label.

- #7632
- #7435
- #7014
- #6744
- #6683
- #6619
- #6178
- #6542
- #5986
- #4216
- #3567
- #3566

**Comment by rdstern** [comment]
@lloyddewit I am not sure if this is the right place for the ideas below.  But happy for you to move them if not.

I tried some examples and made many mistakes partly because I was actually trying to adapt a python script and they are quite a bit different.  That gave rise to 2 small points and then a larger suggestion that I'll add to #7632.

a) The examples all used `=` rather than `<- `for assignment.  I don't mind sticking to the <-, but it may be the time to implement = as an alternative (if easy) and there is no down-side.  Or to stick to the <- as a part of good practice?

b) I often got the wrong number of brackets - mainly from copying code in.  Then R-instat was waiting and whatever was done di nothing.  Is there an easy way of clearing the pending code?  If not, could there be a button for that, or something that can do this?

c) When I rund many lines and have an error, it then keeps running subsequent lines and they all give further errors.  There is a mini-version of this in the ordinary code as well that always gives 2 error messages.  Could we avoid running the remaining code after an error?

**Comment by lloyddewit** [comment]
@rdstern Thanks for the feedback!

> a) The examples all used `=` rather than `<- `for assignment. I don't mind sticking to the <-, but it may be the time to implement = as an alternative (if easy) and there is no down-side. Or to stick to the <- as a part of good practice?

In R, `<-` and `=` have slightly different meanings (scope related) and '<-' is generally preferred (with `=` recommended only for function parameters). However, it is legal R to assign variables with `=` and some programmers prefer it. So R-Instat should accept `=`. This is a known problem and is already covered in one of the existing issues (can't remember which one). To make it explicit, I added a new checkbox above.

> b) I often got the wrong number of brackets - mainly from copying code in. Then R-instat was waiting and whatever was done di nothing. Is there an easy way of clearing the pending code? If not, could there be a button for that, or something that can do this?

This should be covered by the check box above which states 'make 'ctrl-enter' execute the R statement at the cursor (rather than just the line), this should fix the current bugs with multi-line statements'.
Currently, the script window 'ctrl-enter' works line by line until it thinks the statement is complete. This can leave it hanging.

> c) When I run many lines and have an error, it then keeps running subsequent lines and they all give further errors. There is a mini-version of this in the ordinary code as well that always gives 2 error messages. Could we avoid running the remaining code after an error?

R is translated, rather than compiled. If we execute a multiple-statement R script, then the R environment will execute the script statement by statement. It will not stop if there is an error. I tested this in RStudio, and RStudio behaves the same.
I think it would be hard to change this behaviour. Is it a high priority?

**Comment by rdstern** [comment]
@lloyddewit the multiple error feature is not at all a high priority.  

I wonder what happens in RStudio?  My impression is that it gives multiple errors, but all at the same time, in the output window.  We have them through pop-ups and have to click each time and then get the new one.  If I am right with the different behavior, then I wonder if the error statement could have an abort option as well as a click to continue?

**Comment by rdstern** [comment]
@lilyclements you recently mentioned a new feature in RStudio script writing that you really like.  I wonder what it was, and whether it could easily be included into R-Instat scripts?

**Comment by lilyclements** [comment]
@rdstern yes, so I have spoken to @lloyddewit about this - but the feature is a new one in RStudio, where the code is highlights as a colour if you put a colour in quotes.
 E.g.

![image](https://user-images.githubusercontent.com/21180424/224065895-7d028a5e-6ee7-41c9-9008-a51ebffb249b.png)

**Comment by lloyddewit** [comment]
@lilyclements thanks for this, I added the feature to the list in the comment description.
(I already did this last week but I must have forgotten to save the change, happy you reminded me)

**Comment by rdstern** [comment]
@lloyddewit I added another ambitious point, perhaps to your list at the top of this issue:  I copy it here and then add another one!

a) One big failing in R-Instat is that when running a command from a dialog there is no way of stopping the running without ctrl alt delete, which is drastic!

Might we be able to stop when running from a script? This is [discussed here](https://stackoverflow.com/questions/69358524/what-function-to-interrupt-the-running-r-script)

Maybe I am dreaming too much!

But the reason that a script only solution could be very useful is that I usually know which dialogs could be risky in terms of taking a long time. So I could set them up in the dialog and then run them from the script window! One I am always using as a test is the factorize2 on the calculator integer keyboard!

Of course I think Chris may be investigating whether something can be done from the dialogs. But this feature would be useful anyway. 

b) Having a list of past used scripts.

When using RStudio I have the past scripts as buttons at the top.  Could we have a similar list in R-Instat.
And we have a very discrete way we could provide the list!  Could it be added to the down arrow which is next to the Log/Script window.  There is just one entry there currently, and it just says whether the window is open or not.  We might keep that, but underneath is the recent scripts we have opened and saved.
 
I got to this recently when I was sent a set of data, in Excel and a script - so with an R extension.  I used File > Open for both and it was equally simple.  The data went to a data frame and the script went to the script window.   I saved my own copy of the data as an RDS file and the code as an R file. 

Now I returned today and the RDS file is available and in 3 places!  It is on the front screen and is on the File menu and is on the toolbar drop-down.  But my R file is nowhere?  

So the initial opening is the same for data and code.  But then it is different.  We could, of course, save the R file in the list of past files, just like a data file, so that's another option.  I am thinking that we are less obtrusive about the R files - but just as convenient.  (I am also asking if the liobrary icon on the toolbar could have its drop-down, because we currently don't save a copy of those either.  Though that's slightly different, because I anticipate that list just from opening a library file, not from saving a special copy.)

**Comment by lloyddewit** [comment]
@rdstern thank you for the well-explained suggestions. To ensure these suggestions are tracked I opened issues #8486 and #8487, and added these issues to the main list at the start of this PR.
